security/pkg/pki/ca/ca.go

```diff
@@ -210,23 +226,85 @@ func NewSelfSignedIstioCAOptions(ctx context.Context,
 	return caOpts, err
 }
 
-func loadSelfSignedCaSecret(client corev1.CoreV1Interface, namespace string, caCertName string, rootCertFile string, caOpts *IstioCAOptions) error {
+// loadCASecrets loads the original istio-ca-secret Secret from K8S.
+// To simplify - this secret should be used for both self-generated and user defined intermediary certs.
+// TODO: rename to loadRootsFromSecret.
+func loadCASecrets(client corev1.CoreV1Interface, namespace string, caCertName string, rootCertFile string, caOpts *IstioCAOptions) error {
 	caSecret, err := client.Secrets(namespace).Get(context.TODO(), caCertName, metav1.GetOptions{})
 	if err == nil {
-		pkiCaLog.Infof("Load signing key and cert from existing secret %s/%s", caSecret.Namespace, caSecret.Name)
-		rootCerts, err := util.AppendRootCerts(caSecret.Data[CACertFile], rootCertFile)
+		// Original simple case - self-generate root, no intermediaries.
+		rootCertData := caSecret.Data[CACertFile]
+
+		// The file is typically the mounted ca - if the roots are present, they are added.
+		// Not sure if this is intentional or accidental.
+		rootCerts, err := util.AppendRootCerts(rootCertData, rootCertFile)
 		if err != nil {
 			return fmt.Errorf("failed to append root certificates (%v)", err)
 		}
-		if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(caSecret.Data[CACertFile],
-			caSecret.Data[CAPrivateKeyFile], nil, rootCerts); err != nil {
+
+		privData := caSecret.Data[CAPrivateKeyFile]
+		crtData := caSecret.Data[CACertFile]
+
+		if caOpts.KeyCertBundle, err = util.NewVerifiedKeyCertBundleFromPem(crtData,
+			privData, nil, rootCerts); err != nil {
 			return fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
 		}
-		pkiCaLog.Infof("Using existing public key: %v", string(rootCerts))
+
 	}
 	return err
 }
 
+func loadCacertSecret(client corev1.CoreV1Interface, namespace string, caCertName string, caOpts *IstioCAOptions) error {
+	caSecret, err := client.Secrets(namespace).Get(context.TODO(), caCertName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	// Detect tls.key, load both ca-cert.pem and root-cert.pem
+	rootCertData := caSecret.Data[TLSSecretRootCertFile]
+	privData := caSecret.Data[TLSSecretCAPrivateKeyFile]
+	crtData := caSecret.Data[TLSSecretCACertFile]
+
+	if rootCertData != nil {
+		rootCertData = caSecret.Data[CACertFile]
+		rootCertData2 := caSecret.Data[CertChainFile]
+		if rootCertData2 != nil {
+			rootCertData = append(rootCertData, rootCertData2...)
+		}
+		privData = caSecret.Data[CAPrivateKeyFile]
+		crtData = caSecret.Data[CACertFile]
+	}
```

Comment:
Could you explain this code block in a comment?

Reply:
Will do - unless I just remove this code.